src/main/java/org/spongepowered/api/resource/Resource.java

```diff
@@ -0,0 +1,178 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.resource;
+
+import com.google.common.io.ByteStreams;
+import com.google.common.io.CharStreams;
+import com.google.common.io.MoreFiles;
+import org.spongepowered.api.data.DataView;
+import org.spongepowered.api.data.persistence.DataFormat;
+import org.spongepowered.api.data.persistence.InvalidDataFormatException;
+
+import java.io.BufferedReader;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.nio.charset.Charset;
+import java.nio.file.OpenOption;
+import java.nio.file.Path;
+import java.util.Optional;
+import java.util.stream.Stream;
+import javax.annotation.WillNotClose;
+
+/**
+ * A resource can represent any kind of loaded data. It can be a file on the
+ * filesystem, a network location, or even generated at runtime. Use
+ * {@link #getInputStream()} to load the data held by a resource.
+ *
+ * <p>Resource extends Closeable, so don't forget to close your streams either
+ * with a try-with-resources or a close() inside a try-finally block. e.g.</p>
+ * <pre>
+ *     try (Resource res = resourceManager.getResource(path)) {
+ *         logger.info(res.readString(DefaultCharsets.UTF_8));
+ *     } catch (IOException e) {
+ *         logger.warn("Failed to load resource: {}", path, e);
+ *     }
+ * </pre>
+ */
+public interface Resource extends Closeable {
+
+    /**
+     * Returns a new {@link InputStream} of this resource. A new input stream
+     * should be created each time this method is called.
+     *
+     * @return A new input stream
+     */
+    InputStream getInputStream();
+
+    /**
+     * Gets the metadata for this resource.
+     *
+     * <p>The metadata file has the same name as this resource, but has
+     * {@code .mcmeta} appended to the end.</p>
+     *
+     * <p>For example: the metadata for the resource
+     * {@code minecraft:textures/blocks/water_flow.png} would be located at
+     * {@code minecraft:textures/blocks/water_flow.png.mcmeta}</p>
+     *
+     * @return The metadata or {@link Optional#empty() empty} if it doesn't exist.
+     * @see <a href=http://minecraft.gamepedia.com/Resource_pack#Contents> Minecraft Wiki/Resource Packs
+     */
+    Optional<DataView> getMetadata();
+
+    /**
+     * Gets the path of this resource.
+     *
+     * @return The path
+     */
+    ResourcePath getPath();
+
+    /**
+     * Gets the pack name which owns this resource.
+     *
+     * @return The parent pack.
+     */
+    String getPackName();
+
+    /**
+     * Gets a reader for this resource using the given {@link Charset}.
+     *
+     * @param charset The charset
+     * @return A new reader
+     */
+    default BufferedReader getReader(Charset charset) {
```

Comment:
`createReader`/`makeReader`/`newReader` as `get` implies this is an intrinsic property.